rasterio/_io.pyx

```diff
@@ -200,6 +202,11 @@ cdef class DatasetReaderBase(DatasetBase):
 
         cdef GDALRasterBandH band = NULL
 
+        # Reading from a dataset opened in "w" will be deprecated.
+        if self.mode == "w":
+            warnings.warn(
+                "Reading from datasets opened in 'w' mode will not be allowed in the next version. Use 'w+' mode instead.", RasterioDeprecationWarning)
```

Comment:
This is a really long line.
Might be better to make it fit on multiple line so flake8 doesn't yell at you ;-)